WebCore::IDBResourceIdentifier's m_resourceNumber should be an ObjectIdentifier

sheeparegreat opened the pull request:
#### 4510c4321cc95439a76be611a37e8c023772bd5c
<pre>
WebCore::IDBResourceIdentifier&apos;s m_resourceNumber should be an ObjectIdentifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=289201">https://bugs.webkit.org/show_bug.cgi?id=289201</a>
<a href="https://rdar.apple.com/146101559">rdar://146101559</a>

Reviewed by Sihui Liu.

* Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.cpp:
(WebCore::IDBResourceIdentifier::IDBResourceIdentifier):
(WebCore::IDBResourceIdentifier::loggingString const):
* Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.h:
(WebCore::IDBResourceIdentifierHashTraits::constructDeletedValue):
(WebCore::IDBResourceIdentifierHashTraits::isDeletedValue):
(): Deleted.
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.cpp:
(WebKit::dictionaryForWebKitSecureCodingType):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/291789@main">https://commits.webkit.org/291789@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/559b792c32d21eb2bfa5968fbf39cd0d82d11c29

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93879 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13462 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98885 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44406 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95929 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13760 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21895 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71639 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29001 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96881 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10216 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84816 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51974 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/93374 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9897 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2452 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43722 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80162 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2536 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100926 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20931 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15262 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80655 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21183 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80776 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80003 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19955 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24551 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1912 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14088 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20915 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26093 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20602 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24062 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22343 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.cpp:
```diff
@@ -34,29 +34,17 @@
-static uint64_t nextClientResourceNumber()
-{
-    static std::atomic<uint64_t> currentNumber(1);
-    return currentNumber += 2;
```
I'm curious why the existing code was using `+= 2` and not a simple increment. Surely there must have been a reason for it and it is now different with an ObjectIdentifier so I worry. I recommend @szewai and @beidson review this to make sure this is OK.

szewai replied:
(@beidson can correct me if I am wrong) This number can be generated on client side (web process) and on server side (networking process) for `IDBTransactionInfo`, so to make sure they don't generate the same number, one only generates odd number and the other only generates even number.

To avoid breakage, this PR should keep the same behavior by creating ObjectIdentifier with next resource number, instead of using generate() function.

sheeparegreat replied:
Updated

RupinMittal replied:
I may be wrong about this--but could we use `AtomicObjectIdentifier` here? It feels like that would be cleaner and less error prone than one process using even numbers and other using odd ones.

RupinMittal replied:
Perhaps @szewai or @beidson could comment if there's any reason we can't use `AtomicObjectIdentifier`.

szewai replied:
What does it mean by "using `AtomicObjectIdentifier` here"?

RupinMittal replied:
Oh I meant that we should use `AtomicObjectIdentifier` for `IDBResourceObjectIdentifier` instead of `ObjectIdentifier`. But I missed that we are already doing that, sorry! Ignore my comment :)

szewai commented on Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.h:
```diff
@@ -44,6 +44,9 @@ class IDBConnectionToClient;
 struct IDBResourceIdentifierHashTraits;
 
+enum class IDBResourceIdentifierResourceNumberType { };
+using IDBResourceIdentifierResourceNumberID = AtomicObjectIdentifier<IDBResourceIdentifierResourceNumberType>;
```
`Number` and `ID` seem to mean the same thing here; we could just call this `IDBResourceObjectIdentifier`

sheeparegreat committed:
WebCore::IDBResourceIdentifier's m_resourceNumber should be an ObjectIdentifier
https://bugs.webkit.org/show_bug.cgi?id=289201
rdar://146101559

Reviewed by Sihui Liu.

* Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.cpp:
(WebCore::IDBResourceIdentifier::IDBResourceIdentifier):
(WebCore::IDBResourceIdentifier::loggingString const):
* Source/WebCore/Modules/indexeddb/shared/IDBResourceIdentifier.h:
(WebCore::IDBResourceIdentifierHashTraits::constructDeletedValue):
(WebCore::IDBResourceIdentifierHashTraits::isDeletedValue):
(): Deleted.
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.cpp:
(WebKit::dictionaryForWebKitSecureCodingType):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/291789@main
